crates/polars-ops/src/chunked_array/strings/split.rs

```diff
@@ -11,6 +11,7 @@ pub fn split_to_struct<'a, F, I>(
     by: &'a StringChunked,
     n: usize,
     op: F,
+    split_exact: bool,
```

Comment:
There should be no need for this extra argument. Use the fact that we pass `n + 1` as an argument for the `exact` functions and `n` otherwise.

Reply:
Thanks for the feedback! I agree that `s.splitn(n + 1, "").skip(1)` would work in most cases, but when `s` is `"foobar"` and n is 7, we will get `["f", "o", "o", "b", "a", "r", ""]` and I think this would require us to add a `filter` to skip the last empty string.

My current solution is:
- for the `split_exact` case, use `split_terminator` so that we do not need to worry about getting an empty string at the end,
- for the `splitn` case, pre-compute the maximum possible number of substrings (i.e. `std::cmp::min(n, s.chars().count()) + 1`) so that the string will not be splitted after the last character.

What do you think about this edge case? Would it be possible to use `s.splitn(n + 1, "").skip(1)` without an expensive `filter` in that case?

Reply:
@orlp If I understand correctly, the length of `arr_iter` is `n`, so the extra empty string at the end will still be consumed in some cases if we do not filter it out. Let me explain it with a more concrete example.

With my current approach, the result of `pl.Series('name', ["abc"]).str.splitn("", 5)` is

```
shape: (1,)
Series: 'name' [struct[5]]
[
        {"a","b","c",null,null}
]
```

But if we use `s.splitn(n + 1, "").skip(1)` without a `filter`, the result would be

```
shape: (1,)
Series: 'name' [struct[5]]
[
        {"a","b","c","",null}
]
```

Or am I missing something here?

Reply:
Ah, yes, I guess we'll have to make our own split iterator that doesn't have this extra empty stuffix:

```rust
struct SplitNChars<'a> {
    s: &'a str,
    n: usize
}

impl<'a> SplitNChars<'a> {
    fn new(s: &'a str, n: usize) -> Self {
        Self { s, n }
    }
}

impl<'a> Iterator for SplitNChars<'a> {
    type Item = &'a str;
    
    fn next(&mut self) -> Option<Self::Item> {
        if self.n >= 2 {
            self.n -= 1;
            let ch = self.s.chars().next()?;
            let first;
            (first, self.s) = self.s.split_at(ch.len_utf8());
            Some(first)
        } else if self.n == 1 && s.len() > 0 {
            self.n -= 1;
            Some(self.s)
        } else {
            None
        }
    }
}
```

So instead of `s.splitn(n + 1, "").skip(1)` you'll use `SplitNChars { s, n }`. This also handles the case we discussed in the linked issue where we want the empty string split by the empty string to be the empty list.

Reply:
Thanks. I updated the code in 273b37a306293af191340b03ce1d1ddb2ebb2e58 to use `SplitNChars`. I added a `keep_remainder` field to the struct in order to make it also applicable to the `split_exact` case.

Reply:
@haocheng6 Why did you add a boolean flag back? Isn't that already handled by the fact that `split_exact` passes `n + 1` and `splitn` passes `n`? Note that that is a different '`n + 1`' than the one we previously discussed.

Reply:
The boolean flag is needed to handle cases like `pl.Series('name', ["aaaaaaa"]).str.split_exact("", 3)` because I think the desired result for that case is `{"a","a","a","a"}` instead of `{"a","a","a","aaaa"}`.

I updated the implementation as suggested in 5bf814ffc15b59fa5433415321d95565f3e49c62.